src/NodeCompiler/CompileNodeToValue.php

```diff
@@ -273,6 +281,30 @@ private function compileDirConstant(Node\Scalar\MagicConst\Dir $node, CompilerCo
         return dirname(realpath($context->getFileName()));
     }
 
+    /**
+     * Compiles magic constant __CLASS__
+     *
+     * @param Node\Scalar\MagicConst\Class_ $node
+     * @param CompilerContext $context
+     * @return string
+     */
+    private function compileClassConstant(Node\Scalar\MagicConst\Class_ $node, CompilerContext $context): string
+    {
+        return $context->hasSelf() ? $context->getSelf()->getName() : '';
```

Comment:
I think it should match PHP's behaviour no matter how strange it may seem or feel.